### PR TITLE
fix(storage): deleting manifests with identical digests

### PR DESCRIPTION
**Pull request by peusebiu**
Suppose we push two identical manifests (sharing same digest) but with different tags, then deleting by digest should throw an error otherwise we end up deleting all image tags (with gc) or dangling references (without gc)

This behaviour is controlled via Authorization, added a new policy action named detectManifestsCollision which enables this behaviour

Signed-off-by: Ramkumar Chinchani <rchincha@cisco.com>
Signed-off-by: Petu Eusebiu <peusebiu@cisco.com>

**What type of PR is this?**

**Which issue does this PR fix**:

**What does this PR do / Why do we need it**:

**If an issue # is not available please add repro steps and logs showing the issue**:

**Testing done on this change**:

**Automation added to e2e**:

**Will this break upgrades or downgrades?**

**Does this PR introduce any user-facing change?**:

```release-note

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/api/routes.go [hunk 1]
```diff
@@ -475,7 +475,7 @@ func (rh *RouteHandler) UpdateManifest(response http.ResponseWriter, request *ht
 			// could be syscall.EMFILE (Err:0x18 too many opened files), etc
 			rh.c.Log.Error().Err(err).Msg("unexpected error: performing cleanup")
 
-			if err = imgStore.DeleteImageManifest(name, reference); err != nil {
+			if err = imgStore.DeleteImageManifest(name, reference, false); err != nil {
```
Comment (rchincha) [inline pkg/api/routes.go:478]: can we pass a "policy" argument here, so as to become future proof?

File: pkg/api/routes.go [hunk 2]
```diff
@@ -521,14 +521,30 @@ func (rh *RouteHandler) DeleteManifest(response http.ResponseWriter, request *ht
...
+	acCtx, err := localCtx.GetAccessControlContext(request.Context())
+	if err != nil {
+		response.WriteHeader(http.StatusInternalServerError)
+
+		return
+	}
+
+	var detectCollision bool
+	if acCtx != nil {
+		detectCollision = acCtx.CanDetectManifestsCollision(name)
```
Comment (rchincha) [inline pkg/api/routes.go:534]: perfect

File: pkg/api/routes.go (hunk 1 above)
Comment (rchincha) [reply pkg/api/routes.go:478]: Let's do this later.

File: pkg/extensions/search/resolver.go [hunk 3]
```diff
@@ -722,40 +721,21 @@ func BuildImageInfo(repo string, tag string, manifestDigest godigest.Digest,
 	return imageInfo
 }
 
-// returns either a user has or not rights on 'repository'.
-func matchesRepo(globPatterns map[string]bool, repository string) bool {
```
Comment (rchincha) [inline pkg/extensions/search/resolver.go:726]: Why is this part removed?

Comment (peusebiu) [reply pkg/extensions/search/resolver.go:726]: I moved it here: https://github.com/project-zot/zot/pull/951/files#diff-78f3d35f6fab9bf5af09ca967e0e55ec0ccab9cd7ab67fe5dfce6f51ad59f7a6R46
to be more explicit.

File: pkg/api/authz.go [hunk 4]
```diff
@@ -18,11 +18,13 @@ import (
...
 
 const (
-	// actions.
+	// method actions.
 	CREATE = "create"
 	READ   = "read"
 	UPDATE = "update"
 	DELETE = "delete"
+	// behaviour actions.
+	DETECT_MANIFESTS_COLLISION = "detectManifestsCollision" //nolint: stylecheck,revive
```
[the remark this answers is not in the document]
Comment (peusebiu) [reply pkg/api/authz.go:27]: ok, will change

File: pkg/api/authz.go [hunk 5]
```diff
@@ -18,11 +18,13 @@ import (
 )
 
 const (
-	// actions.
+	// method actions.
 	CREATE = "create"
```
Comment (rchincha) [inline pkg/api/authz.go:22]: @peusebiu change these to golang camelcase. These are not special, so no reason to do it this way.

File: pkg/requestcontext/context.go [hunk 6]
```diff
@@ -12,7 +19,54 @@ func GetContextKey() *Key {
...
+	if authCtx := ctx.Value(authzCtxKey); authCtx != nil {
+		acCtx, ok := authCtx.(AccessControlContext)
+		if !ok {
+			return nil, errors.ErrBadType
+		}
+
+		return &acCtx, nil
+	}
+
+	//nolint: nilnil
```
Comment (rchincha) [inline pkg/requestcontext/context.go:41]: put this coment on the same line

File: pkg/api/authz.go (hunk 5 above)
Comment (rchincha) [reply pkg/api/authz.go:22]: also remove the nolint

File: pkg/api/authz.go [hunk 7]
```diff
@@ -18,11 +18,13 @@ import (
...
-	READ   = "read"
-	UPDATE = "update"
-	DELETE = "delete"
+	// method actions.
+	Create = "create"
+	Read   = "read"
+	Update = "update"
+	Delete = "delete"
+	// behaviour actions.
+	DetectManifestCollission = "detectManifestCollision"
```
Comment (rchincha) [inline pkg/api/authz.go:27]: "Collision"

File: examples/README.md [hunk 8]
```diff
@@ -191,7 +194,13 @@ because it will be longer. So that's why we have the option to specify an adminP
 
 Basically '**' means repositories not matched by any other per-repository policy.
 
-create/update/delete can not be used without 'read' action, make sure read is always included in policies!
+Actions list:
```
Comment (rchincha) [inline examples/README.md:197]: ## Method-based
...

## Behavior-based
...

**Review by rchincha — COMMENTED** [review]
Add a bats test case so we know how external tools will behave when they hit this policy.

"skopeo delete should fail"
"regctl"

**Review by rchincha — APPROVED** [review]
lgtm
